src/main/java/crawlercommons/sitemaps/SiteMapParser.java

```diff
@@ -436,18 +445,18 @@ protected SiteMap parseXmlSitemap(URL sitemapUrl, Document doc) {
         SiteMap sitemap = new SiteMap(sitemapUrl);
         sitemap.setType(SitemapType.XML);
 
-        NodeList list = doc.getElementsByTagName("url");
+        NodeList list = doc.getElementsByTagNameNS(Namespace.SITEMAP, "url");
```

Comment:
what would happen here if the namespace is not set explicitly in the XML? Why not do as above and just match based on the localname no matter what the NS is?

Reply:
@jnioche i believe there would be no match without specifying the namespace.  Yes, there would be risk in breaking existing sitemaps that do not specify the namespace.  I was more forgiving about RSS/Atom, since they optionally allow namespaces.

But i believe the Sitemap XML spec requires a namespace...
https://www.sitemaps.org/protocol.html
https://support.google.com/webmasters/answer/183669#errors

And this validator fails when the namespace declaration is missing...
http://freetools.webmasterworld.com/tools/site-validator/

If you match any namespace, then you also risk incorrectly processing an element belonging to another namespace (example, i can imaging that <url> may exist in another mixed-in namespace).

So the decision seems to boil down to whether you wish to support invalid sitemap XML's that do not include the namespace decl.  If that is a common case, then i can see merit, but my impression is that google would reject a Sitemap XML without the ns?  

If you do go forward with this Pull and require the namespace, then it may be a good idea to fail-fast and report an error if the ns is missing?

Reply:
Hi @grokwich,

I've tried your pull request on a set of few 10,000 sitemaps using the SAX parser (see [1](https://groups.google.com/forum/#!topic/crawler-commons/pOLsCVwRsxY)). There is a significant impact: the number of extracted URLs drops from 193 million to 188 million. Afaics, that's (mostly) not caused by absent namespace declarations but by a namespace URL not equal to http://www.sitemaps.org/schemas/sitemap/0.9. E.g., in [this sitemap.xml](http://bvu.edu.vn/sitemap.xml):
```
<urlset xmlns="http://www.google.com/schemas/sitemap/0.9"
        xmlns:xhtml="http://www.w3.org/1999/xhtml">
```
It could be just an outdated location or version (like https://www.google.com/schemas/sitemap/0.84/). But is this really a reason to make the parser fail? What happens if a new sitemap version is released?

I agree that the parser should support namespaces but at least for the SAX parser robustness is most important, with the aim to extract URLs even if the sitemap is not valid. A validation mode or parser would be nice to have but it should be optional as it's just a different use case. If used for a crawler robustness and fault-tolerance matters.

Reply:
Hi @sebastian-nagel.

193M URL's is quite the test-bed:-)

Google Search Console has a sitemap testing tool.  When i test with no namespace decl i get a failure...
![image](https://user-images.githubusercontent.com/4251028/30502346-db75d022-9a33-11e7-9956-8ca865bc268b.png)

Also if i provide a bad namespace URL i get a failure...
![image](https://user-images.githubusercontent.com/4251028/30502369-f20e4b5c-9a33-11e7-962b-26e55925a17e.png)

But interestingly if i test using the URL examples you provided, then the test passes.  I guess this makes sense since the independent sitemap spec evolved from google's old internal sitemap format.  They don't seem to document those legacy namespaces, so i assume they are deprecated but still supported for legacy clients.

I also tried a few 3'rd party validators, and they reject the Google legacy URL's.

http://tools.seochat.com/tools/site-validator
http://freetools.webmasterworld.com/tools/site-validator

The spec is clear that the namespace is required...
https://www.sitemaps.org/protocol.html

Here is also an excerpt from a Google doc to troubleshoot failed processing, where they state that the namespace must exist and have the exact URL specified (expand the "Complete error list" at the bottom of the following page)...

https://support.google.com/webmasters/answer/183669#error-list

Here is the quote...
```
Unsupported format
Your sitemap is not in a supported format. Sitemaps must be in XML format, and use the correct header.
Common XML mistakes:

Your sitemap must use the correct header. For example, if your sitemap contains video information, it would have the following header:
<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">

The namespace in the header must be "http://www.sitemaps.org/schemas/sitemap/0.9" (not .9).
All XML attributes must enclosed in either single quotes (') or double quotes (") and those quotes must be straight, not curly. Word-processing programs such as Microsoft Word can insert curly quotes.
```
Another consideration is when a sitemap adds additional namespaces.  If one of those includes elements names that overlap with element names from sitemap XML, then you run the risk of incorrectly processing an element from another namespace.

For example, here is a case where you would incorrectly <foobar:url>...
```XML
<urlset 
    xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
    xmlns:foobar="http://www.foobar.com/schemas/foobar">

   <url>
   ...
   </url>
   
   <foobar:url>
   ...
   </foobar:url>

   ....
<urlset> 
```
As for versioning, if a new namespace is introduced for sitemap (unlikely), then presumably new capability is being added that will change the way a client should process the doc.  In that case the sitemap parsers would have to be enhanced to support processing the new version, and clients should have to specify which version to parse with (default being the old version).

:warning: *All that being said*, i completely understand the concern of breaking existing (invalid) behavior.  One middle-ground would be to make the default wild-card, but allow the parser to be configured as strict (to avoid namespace clashes).

Whatever pragmatic decision you guys make is good with me.  Thx.

Reply:
Hi @grokwich,

First off, many thanks for the very detailed response & testing! After dealing with all of the fun issues with invalid robot.txt files found in the wild, I appreciate your research and the time spent documenting things.

My two cents - I think your middle-ground solution is a good way to handle the two conflicting goals.

Reply:
+1 to the middle-ground approach
Thanks for the testing and commenting